MoreLinq/Segment.cs

```diff
@@ -76,44 +76,41 @@ public static IEnumerable<IEnumerable<T>> Segment<T>(this IEnumerable<T> source,
 
             return _(); IEnumerable<IEnumerable<T>> _()
             {
-                var index = -1;
                 using var iter = source.GetEnumerator();
 
-                var segment = new List<T>();
-                var prevItem = default(T);
+                // early break
+                if (!iter.MoveNext())
+                    yield break;
 
                 // ensure that the first item is always part
                 // of the first segment. This is an intentional
                 // behavior. Segmentation always begins with
                 // the second element in the sequence.
-                if (iter.MoveNext())
-                {
-                    ++index;
-                    segment.Add(iter.Current);
-                    prevItem = iter.Current;
-                }
+                var prevItem = iter.Current;
+                var segment = new List<T> { iter.Current };
 
-                while (iter.MoveNext())
+                for (var index = 1; iter.MoveNext(); index++)
                 {
-                    ++index;
                     // check if the item represents the start of a new segment
                     var isNewSegment = newSegmentPredicate(iter.Current, prevItem, index);
                     prevItem = iter.Current;
```

Comment:
If this (`prevItem = iter.Current;`) is moved to the bottom of the `for` loop (which seems more natural) then `isNewSegment` can be removed and its assigned expression in-lined into the condition for the following `if`, on line 98. It makes the overall code then much simpler to read and reason about.